coda_v2_python_client/firebase_client_wrapper.py

```diff
@@ -165,3 +199,74 @@ def get_message(self, dataset_id, message_id):
+                                   field and where the LastUpdated field is later than last_updated_after. Defaults to None.
+        :type last_updated_after: datetime, optional
+        :return: Messages in this dataset, filtered by 'LastUpdated' timestamp if requested.
+        :rtype: list of dict
+        """
+        segment_count = self.get_segment_count(dataset_id)
+        if segment_count is None or segment_count == 1:
+            return self.get_segment_messages(dataset_id, last_updated_after)
+        else:
+            # Get the messages for each segment
+            messages_by_segment = dict()  # of segment id -> list of message
+            for segment_index in range(1, segment_count + 1):
+                segment_id = self.id_for_segment(dataset_id, segment_index)
+                messages_by_segment[segment_id] = self.get_segment_messages(segment_id, last_updated_after)
+
+            # Search the fetched segments for the most and least recently updated timestamps in all the segments downloaded
+            # above.
+            dataset_first_updated = None
+            dataset_last_updated = None
+            for segment_messages in messages_by_segment.values():
+                for msg in segment_messages:
+                    msg = msg.to_dict()
+                    if "LastUpdated" in msg:
+                        if dataset_last_updated is None or msg["LastUpdated"] > dataset_last_updated:
+                            dataset_last_updated = msg["LastUpdated"]
+                        if dataset_first_updated is None or msg["LastUpdated"] < dataset_first_updated:
+                            dataset_first_updated = msg["LastUpdated"]
+
+            # Check all the segments for any messages between the latest one we fetched above and the most recently updated
+            # message seen in any segment. If we didn't fetch any new messages for a segment, use the oldest timestamp
+            # across all segments (dataset_first_updated) instead. This is to ensure we don't miss any messages that were
+            # being labelled while we were pulling the separate segments, and is needed to maintain the consistency
+            # guarantees we need for incremental fetch.
+            for segment_id, segment_messages in messages_by_segment.items():
+                segment_last_updated = None
+                for msg in segment_messages:
+                    msg = msg.to_dict()
+                    if "LastUpdated" in msg:
+                        if segment_last_updated is None or msg["LastUpdated"] > segment_last_updated:
+                            segment_last_updated = msg["LastUpdated"]
+
+                if segment_last_updated is None:
+                    segment_last_updated = dataset_first_updated
+
+                if segment_last_updated is not None:
+                    updated_segment_messages = self.get_segment_messages(segment_id, segment_last_updated, dataset_last_updated)  # nopep8
```

Comment:
I think this bit will be slightly clearer if:  
```suggestion
                    updated_segment_messages = self.get_segment_messages(segment_id, last_updated_after=segment_last_updated, last_updated_before=dataset_last_updated)  # nopep8
```